src/daft-scan/src/scan_task_iters/split_parquet_files_by_rowgroup.rs

```diff
@@ -0,0 +1,465 @@
+                        metadata,
+                        ..
+                    } = &mut new_source
+                    {
+                        // Create a new Parquet FileMetaData, only keeping the relevant row groups
+                        let row_group_list =
+                            RowGroupList::from_iter(finalized_rg_idxs.iter().map(|idx| {
+                                (
+                                    *idx,
+                                    self.file_metadata.row_groups.get(idx).unwrap().clone(),
+                                )
+                            }));
+                        let new_metadata = self
+                            .file_metadata
+                            .clone_with_row_groups(num_rows, row_group_list);
+                        *parquet_metadata = Some(Arc::new(new_metadata));
+
+                        // Mutate other necessary metadata
+                        *chunk_spec = Some(ChunkSpec::Parquet(
+                            finalized_rg_idxs.iter().map(|&idx| idx as i64).collect(),
+                        ));
+                        *size_bytes = Some(
+                            finalized_rg_idxs
+                                .iter()
+                                .map(|rg_idx| {
+                                    self.file_metadata
+                                        .row_groups
+                                        .get(rg_idx)
+                                        .unwrap()
+                                        .compressed_size()
+                                })
+                                .sum::<usize>() as u64,
+                        );
+                        *metadata = Some(TableMetadata { length: num_rows });
+                    } else {
+                        unreachable!(
+                            "Parquet file format should only be used with DataSource::File"
+                        );
+                    }
+
+                    // Move state back to Accumulating, and emit data
+                    self.state =
+                        SplitSingleParquetFileByRowGroupsState::Accumulating(*idx, Vec::new(), 0.);
+                    let new_scan_task = ScanTask::new(
+                        vec![new_source],
+                        self.scan_task.file_format_config.clone(),
+                        self.scan_task.schema.clone(),
+                        self.scan_task.storage_config.clone(),
+                        self.scan_task.pushdowns.clone(),
+                        self.scan_task.generated_fields.clone(),
+                    )
+                    .into();
+                    return Some(new_scan_task);
+                }
+            }
+        }
+    }
+}
+
+/// Wraps a ScanTaskRef, indicating whether or not it should be split by Parquet Rowgroups
+enum MaybeSplitScanTaskByParquetRowgroups {
+    NoSplit(ScanTaskRef),
+    Split(SplittableScanTaskRef),
+}
+
+struct SplittableScanTaskRef(ScanTaskRef);
```

Comment:
Not sure what this abstraction provides tbh.

I would do something more like 

```
enum SpiltScanTaskDecision {
  NoSplit(ScanTaskRef),
  Split(ScanTaskRef)
}

impl Deref for SpiltScanTaskDecision { ...

impl SpiltScanTaskDecision {
  fn should_split(&self) -> bool
  
}
```

Reply:
I wanted to implement `get_scan_task_iter` on `SplittableScanTaskRef` to avoid needing to do something like `if x.should_split() { x.get_scan_task_iter() }`

Instead, this felt a bit cleaner because I can do:

```
match x {
    NoSplit() => ...,
    Split(x) => x.get_scan_task_iter(),
}
```